passkey: add krb5 plugin

pbrezina opened the pull request:
The plugin is very similar to idp plugin so I took everything
that can be shared out and converted it into a common API.

Since the communication is now two-way (in idp plugin we only
receive data from ipa-otpd but never send any, now we need to
send data as well) and I didn't find any better attribute in
RADIUS protocol that we could use, all data (both incoming
and outgoing) is send in `Proxy-State`.

1. Initial request (kdc plugin -> ipa_otpd)

```
passkey {
  phase: 0
}
```

2. ipa_otpd reply (ipa_otpd -> kdc plugin -> krb5 client plugin -> krb5_child)

```
passkey {
  phase: 1,
  state: "ipa_otpd state",
  data: {
    domain: "...",
    credential_id_list: ["...", ...],
    user_verification: X,
    cryptographic_challenge: "..."
  }
}
```

3. passkey device reply (krb5_child -> krb5 client plugin -> kdc plugin -> ipa_otpd)

```
passkey {
  phase: 2,
  state: "ipa_otpd state",
  data: {
    credential_id: "...",
    cryptographic_challenge: "...",
    authenticator_data: "...",
    assertion_signature: "..."
  }
}
```

pbrezina commented on src/krb5_plugin/idp/idp_clpreauth.c:
```diff
@@ -209,14 +202,6 @@ sss_idpcl_process(krb5_context context,
-}
-
 krb5_error_code
 clpreauth_idp_initvt(krb5_context context,
```
False positive.

pbrezina commented on src/krb5_plugin/passkey/passkey_clpreauth.c:
```diff
@@ -0,0 +1,185 @@
+}
+
+krb5_error_code
+clpreauth_passkey_initvt(krb5_context context,
```
False positive.

pbrezina commented on src/krb5_plugin/passkey/passkey_kdcpreauth.c:
```diff
@@ -0,0 +1,429 @@
+}
+
+krb5_error_code
+kdcpreauth_passkey_initvt(krb5_context kctx,
```
False positive.

pbrezina commented:
Covscan yields errors around `sss_string_array_free`. It looks like false positive. Do anyone see some issue there?

ikerexxe commented:
We are enabling the compilation of the passkey feature with the `BUILD_PASSKEY` configuration flag. [Example](https://github.com/SSSD/sssd/blob/master/Makefile.am#L191)

> Covscan yields errors around `sss_string_array_free`. It looks like false positive. Do anyone see some issue there?

It seems like the tools isn't working properly.

sumit-bose commented on src/krb5_plugin/passkey/passkey_utils.c:
```diff
@@ -0,0 +1,556 @@
+        return NULL;
+    }
+
+    jroot = json_pack("{s:i, s:s?, s:o?}",
```
It looks like `json_unpack()` cannot handle the JSON NULL value only if read explicitly with `n`. So I would suggest to use `{s:i, s:s*, s:o*}` which would not add the optional attributes at all.

sumit-bose commented on src/krb5_plugin/passkey/passkey_utils.c:
```diff
@@ -0,0 +1,556 @@
+sss_passkey_message_decode_padata(krb5_pa_data *padata)
+{
+    return sss_radius_decode_padata(
+        (sss_radius_message_decode_fn)sss_passkey_message_from_json, padata);
```
I think `sss_passkey_message_decode` would be better here so that the `passkey ` prefix, which you mentioned in the description for the message, is removed.

sumit-bose commented on src/krb5_plugin/common/radius_kdcpreauth.c:
```diff
@@ -0,0 +1,609 @@
+    data.length = strlen(value) + 1;
+
+    ret = sss_radiuskdc_put_complete_attr(attrset,
+                                          krad_code_name2num(attr),
```
`krad_attr_name2num()` is needed here to translate attribute names like `Proxy-State`.

sumit-bose commented on src/krb5_plugin/passkey/passkey_utils.c:
```diff
@@ -0,0 +1,556 @@
+        goto done;
+    }
+
+    ret = json_unpack(jroot, "{s?:o}", "indicators", &jindicators);
```
FreeIPA puts all indicator strings in `[...]`, see https://pagure.io/freeipa/blob/master/f/daemons/ipa-kdb/ipa_kdb_principals.c#_619, so `"[{s?:o}]"` should be used here as well.

pbrezina replied:
otp is required to be array by Kerberos since there is support for multiple token.

For idp, we wanted to use array as well since there may be multiple idp mechanisms in the future.

I think for passkey, it is sufficient to be a single value since it implements a specific protocol.

sumit-bose replied:
Hi,

if you prefer it this way you have to coordinate with @flo-renaud and @abbra since currently the array syntax is used at https://github.com/flo-renaud/freeipa/blob/passkey/daemons/ipa-kdb/ipa_kdb_principals.c#L647.

bye,
Sumit

pbrezina replied:
FYI @flo-renaud @abbra what is your opinion on this?

abbra replied:
I'm with @sumit-bose. Let us keep the syntax the same for all those strings on IPA side:
```
    const krb5_octet rad_string[] = "otp\0[{\"indicators\": [\"radius\"]}]";
    const krb5_octet otp_string[] = "otp\0[{\"indicators\": [\"otp\"]}]";
    const krb5_octet idp_string[] = "idp\0[{\"type\":\"oauth2\",\"indicators\": [\"idp\"]}]";
    const krb5_octet passkey_string[] = "passkey\0[{\"indicators\": [\"passkey\"]}]";
```

abbra replied:
A bit of reasoning: I have plans on extending the speicific indicators in future to include details that might be useful for the applications which would authorize access based on the authentication indicators. This is to not exclude level of assurance (RFC 6711) and similar methods specifically mentioned in the authentication indicator's RFC 8129.

pbrezina replied:
Ok, I changed it.

sumit-bose commented:
Hi,

I started to add comments while testing the plugin with my ipa-otpd extensions. I will let you know when I'm done so that you can handle all comments in the single run.

bye,
Sumit

sumit-bose commented on src/krb5_plugin/common/radius_kdcpreauth.c:
```diff
@@ -0,0 +1,609 @@
+    }
+
+    /* Unexpected padata. Return error. */
+    if (padata->length != 0) {
```
Hi,

this makes sense for IdP where we do not expect a reply. But for passkey there will be padata in the response. So I think this check should be removed here and maybe added back to an IdP specific wrapper when then calls `sss_radiuskdc_return_padata()` for additional checks.

bye,
Sumit

sumit-bose commented on src/krb5_plugin/passkey/passkey_kdcpreauth.c:
```diff
@@ -0,0 +1,429 @@
+        goto done;
+    }
+
+    if (pa->pa_type != SSSD_PASSKEY_PADATA || pa->length != 0) {
```
Hi,

we expect padata here, so the check should be `pa->length == 0`.

bye,
Sumit

sumit-bose commented on src/krb5_plugin/passkey/passkey_kdcpreauth.c:
```diff
@@ -0,0 +1,429 @@
+    }
+
+    if (message->phase != SSS_PASSKEY_PHASE_REPLY
+        || strlen(message->state) != rstate.length
```
Hi,

`rstate.length` includes the trailing `\0` byte, so the check should be `strlen(message->state) + 1 != rstate.length`.

bye,
Sumit

sumit-bose commented on src/krb5_plugin/passkey/passkey_kdcpreauth.c:
```diff
@@ -0,0 +1,429 @@
+    }
+
+    cb->free_string(kctx, rock, configstr);
+    sss_passkeykdc_config_free(config);
```
Hi,

`config->passkey->indicators`  is used in `sss_radiuskdc_verify_done()` which is currently called after the free here. So either the indicators should be copied in `sss_passkeykdc_verify_send()` or the whole `config` should be passed down to `sss_radiuskdc_verify_done()`  and free-d there.

bye,
Sumit

sumit-bose commented on src/krb5_plugin/passkey/passkey_kdcpreauth.c:
```diff
@@ -0,0 +1,429 @@
+    vt->verify = sss_passkeykdc_verify;
+    vt->return_padata = sss_radiuskdc_return_padata;
+
+    com_err(SSSD_PASSKEY_PLUGIN, 0, "Loaded");
```
Hi,

there are now three modules (certauth, IdP and passkey pre-auth) which just print `Loaded` in the logs:
```
Jan 02 12:38:47 master.ipa.vm krb5kdc[479561](info): Loaded
Jan 02 12:38:47 master.ipa.vm krb5kdc[479561](info): Loaded
Jan 02 12:38:47 master.ipa.vm krb5kdc[479561](info): Loaded
```
I think it would help to add the module's name here (and for the other modules) as well.

bye,
Sumit

pbrezina replied:
Changed for IdP and passkey. I was not able to find other "Loaded" string in SSSD nor in IPA.

sumit-bose commented:
Hi,

I'm done with testing and published the ipa-otpd patches with some hints how to test at https://github.com/flo-renaud/freeipa/pull/32.

Besides in in-line comments I have two questions about the proposed JSON protocol data. First, I think the state shouldn't be send to the client. The state is basically the cryptographic challenge and this should not leave the KDC because if the KDC is using the random string returned from the client this can be used for replay attacks. My suggestion would be to extract the cryptographic challenge on the KDC and store it in the cookie before sending the challenge to the client and then compare the cryptographic challenge from the reply with the cookie data.

My second question is about the allowed value for `user_verification`. `passkey_child` has `true` or `false` while `ipa passkeyconfig-mod` allows `on`, `off` and `default`. I'm not sure how to handle `default`. Maybe @ikerexxe or @flo-renaud have comments here as well.

bye,
Sumit

ikerexxe commented:
> Hi,
> 
> I'm done with testing and published the ipa-otpd patches with some hints how to test at [flo-renaud/freeipa#32](https://github.com/flo-renaud/freeipa/pull/32).
> 
> Besides in in-line comments I have two questions about the proposed JSON protocol data. First, I think the state shouldn't be send to the client. The state is basically the cryptographic challenge and this should not leave the KDC because if the KDC is using the random string returned from the client this can be used for replay attacks. My suggestion would be to extract the cryptographic challenge on the KDC and store it in the cookie before sending the challenge to the client and then compare the cryptographic challenge from the reply with the cookie data.
> 
> My second question is about the allowed value for `user_verification`. `passkey_child` has `true` or `false` while `ipa passkeyconfig-mod` allows `on`, `off` and `default`. I'm not sure how to handle `default`. Maybe @ikerexxe or @flo-renaud have comments here as well.
> 
> bye, Sumit

I'm bringing these two topics to our next meeting so that we can discuss them there.

pbrezina commented:
>  I'm done with testing and published the ipa-otpd patches with some hints how to test at https://github.com/flo-renaud/freeipa/pull/32.

Thank you. I addressed the review comments.

> Besides in in-line comments I have two questions about the proposed JSON protocol data. First, I think the state shouldn't be send to the client. The state is basically the cryptographic challenge and this should not leave the KDC because if the KDC is using the random string returned from the client this can be used for replay attacks. My suggestion would be to extract the cryptographic challenge on the KDC and store it in the cookie before sending the challenge to the client and then compare the cryptographic challenge from the reply with the cookie data.

By state, do you mean the crypto challenge attribute from `struct sss_passkey_challenge` or the radius state? Or both?

pbrezina commented:
> I'm bringing these two topics to our next meeting so that we can discuss them there.

Was there any resolution?

ikerexxe commented:
@pbrezina 

> Besides in in-line comments I have two questions about the proposed JSON protocol data. First, I think the state shouldn't be send to the client. The state is basically the cryptographic challenge and this should not leave the KDC because if the KDC is using the random string returned from the client this can be used for replay attacks. My suggestion would be to extract the cryptographic challenge on the KDC and store it in the cookie before sending the challenge to the client and then compare the cryptographic challenge from the reply with the cookie data.

We have to send the cryptographic challenge to the client, but KDC should store it as state to make sure the client returns the same value.

> My second question is about the allowed value for `user_verification`. `passkey_child` has `true` or `false` while `ipa passkeyconfig-mod` allows `on`, `off` and `default`. I'm not sure how to handle `default`. Maybe @ikerexxe or @flo-renaud have comments here as well.

This was discussed on Tuesday's presentation about passkey. We'll change the default to `on/true` to increase the security level.

sumit-bose commented:
> > I'm done with testing and published the ipa-otpd patches with some hints how to test at [flo-renaud/freeipa#32](https://github.com/flo-renaud/freeipa/pull/32).
> 
> Thank you. I addressed the review comments.
> 
> > Besides in in-line comments I have two questions about the proposed JSON protocol data. First, I think the state shouldn't be send to the client. The state is basically the cryptographic challenge and this should not leave the KDC because if the KDC is using the random string returned from the client this can be used for replay attacks. My suggestion would be to extract the cryptographic challenge on the KDC and store it in the cookie before sending the challenge to the client and then compare the cryptographic challenge from the reply with the cookie data.
> 
> By state, do you mean the crypto challenge attribute from `struct sss_passkey_challenge` or the radius state? Or both?

Hi,

both. My idea is that the KDC keeps a copy of the cryptographic challenge and checks if the reply from the client still contains the same and if not just return an error to the clients. This way the KDC can reject replay attacks right away without the need to create a new ipa-otpd process.

bye,
Sumit

pbrezina commented:
Done.

Do we agree on the PA_DATA format or is it still subject to change? If we agree, I will submit request to IANA to register the number.

ikerexxe commented:
> Done.
> 
> Do we agree on the PA_DATA format or is it still subject to change? If we agree, I will submit request to IANA to register the number.

After the discoverable credentials update I don't think we'll change it again:
```
    "passkey": {
        "credential_id": "$credential_id",
        "cyrptographic_challenge": "$cryptographic_challenge",
        "authenticator_data" :"$authenticator_data",
        "assertion_signature": "$assertion_signature",
        "user_id": "$user_id"
    }
```

justin-stephenson commented:
Hi, trying to run `make rpms` with this PR and seeing:
~~~
FAIL: test_krb5_passkey_plugin                                                                                         
==============================                                                                                         
                                                           
[==========] Running 12 test(s).                                                                                       
[ RUN      ] test_sss_passkey_message_encode__null                                                                     
[       OK ] test_sss_passkey_message_encode__null
[ RUN      ] test_sss_passkey_message_encode__invalid                                                                  
[       OK ] test_sss_passkey_message_encode__invalid                                                                  
[ RUN      ] test_sss_passkey_message_encode__init
[  ERROR   ] --- "passkey {"phase":0}" != "passkey {"phase":0,"state":null,"data":null}"
[   LINE   ] --- src/tests/cmocka/test_krb5_passkey_plugin.c:55: error: Failure!       
[  FAILED  ] test_sss_passkey_message_encode__init
[ RUN      ] test_sss_passkey_message_encode__challenge                                                                
[       OK ] test_sss_passkey_message_encode__challenge                                                                
[ RUN      ] test_sss_passkey_message_encode__reply                                                                    
[       OK ] test_sss_passkey_message_encode__reply                                                                    
[ RUN      ] test_sss_passkey_message_decode__null                                                                     
[       OK ] test_sss_passkey_message_decode__null                                                                     
[ RUN      ] test_sss_passkey_message_decode__invalid                                                                  
[       OK ] test_sss_passkey_message_decode__invalid                                                                  
[ RUN      ] test_sss_passkey_message_decode__init
[       OK ] test_sss_passkey_message_decode__init                                                                     
[ RUN      ] test_sss_passkey_message_decode__challenge                                                                
[       OK ] test_sss_passkey_message_decode__challenge
[ RUN      ] test_sss_passkey_message_decode__reply                                                                    
[       OK ] test_sss_passkey_message_decode__reply                                                                    
[ RUN      ] test_sss_passkey_config_init__invalid
[  ERROR   ] --- 0 != 0x16                                                                                             
[   LINE   ] --- src/tests/cmocka/test_krb5_passkey_plugin.c:405: error: Failure!      
[  FAILED  ] test_sss_passkey_config_init__invalid
[ RUN      ] test_sss_passkey_config_init__ok
[  ERROR   ] --- 0x16 != 0
[   LINE   ] --- src/tests/cmocka/test_krb5_passkey_plugin.c:423: error: Failure!
[  FAILED  ] test_sss_passkey_config_init__ok
[==========] 12 test(s) run.
[  PASSED  ] 9 test(s).
[  FAILED  ] 3 test(s), listed below:
[  FAILED  ] test_sss_passkey_message_encode__init
[  FAILED  ] test_sss_passkey_config_init__invalid
[  FAILED  ] test_sss_passkey_config_init__ok

 3 FAILED TEST(S)
FAIL test_krb5_passkey_plugin (exit status: 3)
~~~

pbrezina commented:
I fixed the test and added user_id to reply.

justin-stephenson commented on src/krb5_plugin/passkey/sssd_enable_passkey:
```diff
@@ -0,0 +1,14 @@
+# Enable SSSD OAuth2 Kerberos preauthentication plugins.
+#
+# This will allow you to obtain Kerberos TGT through OAuth2 authentication.
```
and here

justin-stephenson commented on src/krb5_plugin/passkey/sssd_enable_passkey:
```diff
@@ -0,0 +1,14 @@
+#
+# This will allow you to obtain Kerberos TGT through OAuth2 authentication.
+#
+# To disable the OAuth2 plugin, comment out the following lines.
```
and here

ikerexxe commented:
> I fixed the test and added user_id to reply.

Thanks! I updated the [COPR repo](https://copr.fedorainfracloud.org/coprs/ipedrosa/passkey-auth/) with the patches in this PR. While doing that I noticed two things. The first one is that you changed the spec file to include passkey stuff, including a new subpackage. I had already done that in [PR6530](https://github.com/SSSD/sssd/pull/6530). I updated it with your changes, so there's no need to keep them here.

The second thing is that you are compiling everything unconditionally, but we had agreed to add a specific configuration. Please check the `BUILD_PASSKEY` to conditionally build the passkey code.

justin-stephenson commented on src/krb5_plugin/passkey/passkey.h:
```diff
@@ -0,0 +1,97 @@
+
+#define SSSD_PASSKEY_PLUGIN "passkey"
+#define SSSD_PASSKEY_CONFIG "passkey"
+#define SSSD_PASSKEY_PADATA -152 // PA-REDHAT-PASSKEY
```
SSSD_IDP_OAUTH2_PADATA is already 152, does this need to be a different number?

pbrezina replied:
Yes. The number denotes the data format that is send in padata. Since it is different from what we do in oauth2 we need to register a new number. Negative numbers are used for unregistered data but that should not go to production.

pbrezina commented:
> The second thing is that you are compiling everything unconditionally, but we had agreed to add a specific configuration. Please check the `BUILD_PASSKEY` to conditionally build the passkey code.

Fixed.

justin-stephenson commented:
Hi Pavel, I will post some comments in the coming days/week as I am now testing this code in my passkey development work.

justin-stephenson commented:
Plugin and API functions are working well so far in testing.  For the reply I did need to add a passkey_utils function to take the passkey child --get-assert command output, and transform it to the format which the passkey krb5 plugin expects, adding the prefix `passkey {phase, state, data}` parts and returning a struct sss_passkey_message.

The below function converts
~~~
{"credential_id": "xg6rAN18gKM/iI+OFj3cdEs9Fl9kw6MbHt7L+w5ntLk96MVtLUBmDEUSq2gRfD7X4MsDnercMW93xurY0yRyZA==", "cryptographic_challenge": "Gk+87YgAV
jVIXuUY51QrALeSKEltXZHe0Apqil0EQIA=", "authenticator_data": "WCWcAO58pq6rRNqWEXg/P4ZkES9OqDS8L29D9KqffS71WwUAAAFJ", "assertion_signature": "MEUCIDFaoLIj8lrBvAu0xAIs6oomR3SaP+XIn/Qfvx7AIQRMAiEA+zjV9cVvjUiMs6Sgd1fRYwqL/tuJ+xVO4qDBlohj7bw=",
 "user_id": ""}
~~~
To this.
~~~
passkey {"phase":2,"state":"ipa_otpd state","data":{"credential_id":"xg6rAN18gKM/iI+OFj3cdEs9Fl9kw6MbHt7L+w5ntLk96MVtLUBmDEUSq2gRfD7X4MsDnercMW
93xurY0yRyZA==","cryptographic_challenge":"Gk+87YgAVjVIXuUY51QrALeSKEltXZHe0Apqil0EQIA=","authenticator_data":"WCWcAO58pq6rRNqWEXg/P4ZkES9OqDS8L29D9KqffS71WwUAAAFJ","assertion_signature":"MEUCIDFaoLIj8lrBvAu0xAIs6oomR3SaP+XIn/Qfvx7AIQRMAi
EA+zjV9cVvjUiMs6Sgd1fRYwqL/tuJ+xVO4qDBlohj7bw=","user_id":""}}
~~~



~~~
    struct sss_passkey_message *reply_msg;

    phase = SSS_PASSKEY_PHASE_REPLY;
    state = "ipa_otpd state";
    reply_msg = sss_passkey_prefix_json_data(phase, state, reply);

...

+struct sss_passkey_message *
+sss_passkey_prefix_json_data(enum sss_passkey_phase phase,
+                             char *state,
+                             const char *json_str)
+{
+    int ret;
+    json_error_t jret;
+    json_t *jroot;
+    struct sss_passkey_message *message;
+    void *data;
+
+    if (json_str == NULL) {
+        return NULL;
+    }
+
+    jroot = json_loads(json_str, 0, &jret);
+    if (jroot == NULL) {
+        return NULL;
+    }
+
+    data = sss_passkey_reply_from_json_object(jroot);
+    if (data == NULL) {
+        return NULL;
+    }
+
+    message = sss_passkey_message_init(phase, state, data);
+    if (message == NULL && phase == SSS_PASSKEY_PHASE_CHALLENGE) {
+        sss_passkey_challenge_free(data);
+    } else if (message == NULL && phase == SSS_PASSKEY_PHASE_REPLY) {
+        sss_passkey_reply_free(data);
+    }
+
+    json_decref(jroot);
+    return message;
+}


~~~

pbrezina commented:
@justin-stephenson it may be better to add `sss_passkey_reply_decode` to follow the naming convention. This will yield `struct sss_passkey_reply` which you then add to the message as data and encode it with `sss_passkey_message_encode`.

pbrezina commented:
Rebased spec file.

justin-stephenson commented:
> @justin-stephenson it may be better to add `sss_passkey_reply_decode` to follow the naming convention. This will yield `struct sss_passkey_reply` which you then add to the message as data and encode it with `sss_passkey_message_encode`.

IIUC I just need to change the name of the function, and include `sss_passkey_message_encode` at the end so it returns a string, right?

justin-stephenson commented:
> 

Or do you mean `sss_passkey_reply_decode` should return a `struct sss_passkey_reply` and the caller will then carry out the remaining steps (add to message, and call sss_passkey_message_encode)

pbrezina commented:
> > 
> 
> Or do you mean `sss_passkey_reply_decode` should return a `struct sss_passkey_reply` and the caller will then carry out the remaining steps (add to message, and call sss_passkey_message_encode)

I  meant this, but of course you can create the message directly, I'm just not very fond of the name. But we can solve this in your PR.

I'm glad that it works correctly as I did not have any means to test it. I have asked for the PADATA number, so once we have it we can finish this  PR.

pbrezina commented:
We got 153:PA-REDHAT-PASSKEY registered, I just updated `SSSD_PASSKEY_PADATA` to `153` and rebased.

abbra commented:
Is it a typo at https://www.iana.org/assignments/kerberos-parameters/kerberos-parameters.xhtml with PASKEY (one S)? Or you asked for the PA-REDHAT-PASKEY? :)

sumit-bose commented on src/krb5_plugin/passkey/passkey.h:
```diff
@@ -0,0 +1,97 @@
+
+#define SSSD_PASSKEY_PLUGIN "passkey"
+#define SSSD_PASSKEY_CONFIG "passkey"
+#define SSSD_PASSKEY_PADATA 153 // PA-REDHAT-PASSKEY
```
Hi,

just a nitpick, but coding style still says that `//` style comments should be avoided. Otherwise code looks good and is working as expected.

bye,
Sumit

pbrezina replied:
Fixed, thank you.

ikerexxe commented:
Just a general comment for future PRs, I think you should split the test cases into their own function. This way, when one fails you know what's failing exactly. As an extreme example, the [test_sss_passkey_message_encode__challenge()](https://github.com/SSSD/sssd/pull/6628/files#diff-395c70925d974845fd1752fd2fc71078beb13cf06d888fea85c02788918bc55dR64) function contains 9 different test cases. This could be divided in 9 different functions, which would make the reading and maintenance easier. But as I said, this is just a suggestion for future patches.

ikerexxe commented on src/tests/cmocka/test_krb5_passkey_plugin.c:
```diff
@@ -0,0 +1,480 @@
+/*
+    Copyright (C) 2022 Red Hat
```
s/2022/2023

pbrezina commented on src/krb5_plugin/passkey/passkey_clpreauth.c:
```diff
@@ -0,0 +1,185 @@
+    /* Call prompter if we have no answer. We don't really require any answer,
+     * but we need to present a prompt to the user and wait until the user has
+     * finished authentication via an passkey provider. */
+    if (answer == NULL) {
```
I changed the comment. We still have to display the prompt, this is used by `kinit`. It needs to be implemented in `sss_passkeycl_prompt` but I want to leave that up to you or Iker in another PR.

pbrezina commented:
I also opened https://github.com/krb5/krb5/pull/1297

pbrezina commented:
This was merged by accident with Justin patches, unfortunately, he did not have the latest rebase. However it is copyrights and one comment. I opened a new PR for that:
https://github.com/SSSD/sssd/pull/6662
